Send email notification on bad push evolution

La0 opened the pull request:
Closes #625 

The previus classification is retrieved from its indexed artifact, using a new data contract.

I moved the email sending code into taskcluster utils so that method is available to both `classify` and `classify-eval` command.

The email body is pretty simple and would render like this:

```
# Push 157518 evolved from GOOD to BAD

Rev: 46c23a830db14901758da2b09c144d6d1b0e2ad7

## Real failures

- linux64-test/whatever

```

marco-c commented on mozci/util/taskcluster.py:
```diff
@@ -124,3 +125,29 @@ def get_taskcluster_options():
     return options
+
+
+def send_admin_emails(subject, content, emails):
```
Why 'admin'?

La0 replied:
Because I thought we would only have one list of emails in the secret that would be the admins

marco-c commented on mozci/console/commands/push.py:
```diff
@@ -211,6 +223,38 @@ def handle(self) -> None:
                 )
 
+            # Send a notification when some emails are declared in the config
+            emails = config.get("emails")
```
We should use two different configs to store the lists of emails (the receivers of one won't be the same as the receivers of the other).

marco-c commented on mozci/console/commands/push.py:
```diff
@@ -34,6 +35,17 @@
+EMAIL_PUSH_EVOLUTION = """
+# Push {push.id} evolved from {previous.name} to {classification.name}
+
+Rev: {push.rev}
```
Nit: maybe nice adding a link to treeherder

marco-c commented on mozci/console/commands/push.py:
```diff
@@ -211,6 +223,38 @@ def handle(self) -> None:
+                            previous=previous,
+                            classification=classification,
+                            push=push,
+                            real_failures="\n - ".join(regressions.real.keys()),
```
What exactly is this printing? Just the task names or also the group names?

La0 replied:
Just the group name

marco-c replied:
Ok, let's print both the task names and the group names that are failing, so it's easier for the receiver to see it.

We could potentially even give a treeherder link to the exact failure, like this:
https://github.com/mozilla/code-review/blob/9dc5e90e4f011e5b157ef679457b91e38d9bdc37/events/code_review_events/bugbug_utils.py#L394
https://github.com/mozilla/code-review/blob/9dc5e90e4f011e5b157ef679457b91e38d9bdc37/tools/code_review_tools/treeherder.py#L8
(treeherder link can be a follow-up if you want)

marco-c replied:
For example, for the failure browser/extensions/formautofill/test/browser/creditcard/browser.ini on the push a867a89b52a6255e93a568ecc70cf6e0fe2387e0, we can do something like this:

# Real failures
- Group [group1](https://treeherder.mozilla.org/jobs?repo=autoland&tochange=a867a89b52a6255e93a568ecc70cf6e0fe2387e0&test_paths=browser%2Fextensions%2Fformautofill%2Ftest%2Fbrowser%2FcreditCard%2Fbrowser.ini) - Tasks [test-linux1804-64-qr/opt-mochitest-browser-chrome-swr-e10s-7](https://treeherder.mozilla.org/jobs?repo=autoland&revision=a867a89b52a6255e93a568ecc70cf6e0fe2387e0&selectedTaskRun=ehq3v2xQT0CfDvmcJ5UsbQ.0), [test-linux1804-64-qr/opt-mochitest-browser-chrome-spi-nw-e10s-7
](https://treeherder.mozilla.org/jobs?repo=autoland&revision=a867a89b52a6255e93a568ecc70cf6e0fe2387e0&selectedTaskRun=bVo5cDpGR3mPsMslKmZQLQ.0)

In the first link, tochange is a867a89b52a6255e93a568ecc70cf6e0fe2387e0, but actually we should use the push revision where the group actually run (in case it is not the current push like in this case but a following-one). We can start simple and just put the current push revision though.

La0 commented:
I have fixed all suggestions and added a unit for every combination of status evolution. This will showcase the content being sent to email receivers (and possibly change it if you want to add task details...)

:warning: The Taskcluster will need to change towards:

```
[emails]
monitoring = [ ... ]
classifications = [ ... ]
```

EvaBardou commented on tests/test_notify.py:
```diff
@@ -0,0 +1,143 @@
+        status=200,
+    )
+
+    # Run the notificayion code from mozci push classify
```
```suggestion
    # Run the notification code from mozci push classify
```

marco-c reviewed (approved):
LGTM!
